docs/inline.md

```diff
@@ -1,72 +1,103 @@
 # Inline
 
-`inline` is a new [soft modifier][soft-modifier] that guarantees that a definition will be inlined at the point of use.
+Inlining is a common compile-time meta-programming technique for performance optimizations.
+Scala 3 makes several improvements related to inlining:
 
-## Inline values
+1. It introduces `inline` as a [keyword][soft-modifier].
+2. It introduces _inline expressions_.
+3. It guarantees that an _inline definition_ is inlined at the point of usage and an _inline expression_ is always reduced.
 
-The simplest form of inlining comes in the form of known literal constants.
 
-In Scala 2 we would have used a `final val` without a return type for this purpose.
+## Inline constants
 
-```scala
-final val pi = 3.141592653589793
-final val pie = "🥧"
-```
-The compiler operated on a best-effort basis to inline the values.
-
-In Scala 3 we make the first step towards inlining similarly but with stronger guarantees. Firstly, we use the modifier explicitly.
-Secondly, `inline val` *guarantees* that all references to these constants are inlined.
+The simplest form of inlining is to inline constants in programs:
 
 
 ```scala
 inline val pi = 3.141592653589793
 inline val pie = "🥧"
 ```
 
+The usage of the keyword `inline` above *guarantees* that all references to `pi` and `pie` are inlined:
+
 ```scala
 val pi2 = pi + pi // val pi2 = 6.283185307179586
 val pie2 = pie + pie // val pie2 = "🥧🥧"
 ```
 
-As there are no direct references to `pi` and `pi2`, these definitions can be removed from the generated bytecode. There are some exceptions that we will see in [later](#inline-methods).
+In the code above, the references `pi` and `pie` are inlined.
+Then constant folding optimization in the compiler will compute the resulting value `pi2` and `pie2` at _compile-time_.
 
-## Inline Functions
+In Scala 2, we would have used the modifier `final` in the definition that is without a return type:
+
+```scala
+final val pi = 3.141592653589793
+final val pie = "🥧"
+```
 
-We use `inline def` to define a function that will be inlined at a call-site.
+The `final` modifier will ensure that `pi` and `pie` will take a _literal type_.
+Then the constant propagation optimization in the compiler can perform inlining for such definitions.
+However, inlining based constant propagation is _best-effort_ and not guaranteed.
+
```

Comment:
```suggestion
The `final val` inlining is also supported by Scala 3.0 as _best-effort_ inlining for migration purposes.

```

Reply:
I am not sure it's just for migration unless we support const-folded inline expressions. See the example below, and it's commonly used in Dotty.

Reply:
It has nothing to do with constant folding, it is the `final val` that would have a widen type and `inline val` would fully replace those use cases.